File: 30 Days of Code/24 - Day 23 - BST Level-Order Traversal.py
# ========================
#       Information
# ========================

# Direct Link: https://www.hackerrank.com/challenges/30-binary-trees/problem
# Difficulty: Easy
# Max Score: 30
# Language: Python

# ========================
#         Solution
# ========================

import sys

class Node:
    def __init__(self, data):
        self.right = self.left = None
        self.data = data
class Solution:
    def insert(self, root, data):
        if root == None:
            return Node(data)
        else:
            if data <= root.data:
                cur = self.insert(root.left, data)
                root.left = cur
            else:
                cur = self.insert(root.right, data)
                root.right = cur
        return root

    def levelOrder(self, root):
        # Write your code here
        nodes_queue = []
        data_list = []
        temp_node = root

        while temp_node:
            data_list.append(temp_node.data)
            if temp_node.left is not None:
                nodes_queue.append(temp_node.left)
            if temp_node.right is not None:
                nodes_queue.append(temp_node.right)
            if len(nodes_queue) == 0:
                break
            temp_node = nodes_queue.pop(0)

        print(' '.join(list(map(str, data_list))))

T = int(input())
myTree = Solution()
root = None
for i in range(T):
    data = int(input())
    root = myTree.insert(root, data)
myTree.levelOrder(root)

File: 30 Days of Code/C#/24 - Day 23 - BST Level-Order Traversal.cs
// ========================
//       Information
// ========================

// Direct Link: https://www.hackerrank.com/challenges/30-binary-trees/problem
// Difficulty: Easy
// Max Score: 30
// Language: C#

// ========================
//         Solution
// ========================

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Node {
	public Node left,
	right;
	public int data;
	public Node(int data) {
		this.data = data;
		left = right = null;
	}
}
class Solution {
	static void levelOrder(Node root) {
		//Write your code here
		Queue Q = new Queue();
		Q.Enqueue(root);

		while (Q.Count > 0) {
			Node curNode = (Node) Q.Peek();
			if (curNode.left != null) {
				Q.Enqueue(curNode.left);
			}
			if (curNode.right != null) {
				Q.Enqueue(curNode.right);
			}
			Console.Write(curNode.data + " ");
			Q.Dequeue();
		}
	}

	static Node insert(Node root, int data) {
		if (root == null) {
			return new Node(data);
		}
		else {
			Node cur;
			if (data <= root.data) {
				cur = insert(root.left, data);
				root.left = cur;
			}
			else {
				cur = insert(root.right, data);
				root.right = cur;
			}
			return root;
		}
	}

	static void Main(String[] args) {
		Node root = null;

		int T = Int32.Parse(Console.ReadLine());
        
		while (T-->0) {
			int data = Int32.Parse(Console.ReadLine());
			root = insert(root, data);
		}
		levelOrder(root);
	}
}
